src/llmcompressor/modifiers/quantization/calibration.py

```diff
@@ -0,0 +1,251 @@
+import logging
+
+import torch
+from compressed_tensors.quantization import QuantizationStatus, is_attention_module
+from compressed_tensors.quantization.lifecycle.forward import forward_quantize
+from compressed_tensors.quantization.utils import is_kv_cache_quant_scheme
+from compressed_tensors.utils.offload import is_module_offloaded, update_parameter_data
+from torch.nn import Module
+
+from llmcompressor.modifiers.quantization.cache import QuantizedKVParameterCache
+from llmcompressor.observers import Observer
+
+__all__ = [
+    "initialize_observer",
+    "update_weight_zp_scale",
+    "calibrate_input_hook",
+    "calibrate_output_hook",
+    "calibrate_kv_cache_input_hook",
+    "calibrate_kv_cache_output_hook",
+    "set_unset_kv_cache",
+    "freeze_module_quantization",
+]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def initialize_observer(
+    module: Module,
+    base_name: str,
+):
+    """
+    Initialize observer module and attach as submodule.
+    The name of the observer is fetched from the quantization_args.
+    The name is then used to load the observer from the registry and attached
+    to the module. The name of the observer uses the base_name provided.
+
+    :param module: torch.nn.Module that the observer is being attached to
+    :param base_name: str used to name the observer attribute
+
+    """
+
+    arg_name = "weights" if base_name == "weight" else f"{base_name}_activations"
+    quantization_scheme = getattr(module, "quantization_scheme", None)
+    if not quantization_scheme:
+        # no quantization scheme nothing to do
+        return
+
+    # observers have a different lifecycle for kv_cache
+    if is_attention_module(module):
+        return
+
+    quantization_args = getattr(quantization_scheme, arg_name, None)
+    # dont need observers for dynamic
+    if quantization_args and not quantization_args.dynamic:
+        observer = quantization_args.get_observer()
+        observer = Observer.load_from_registry(
+            observer, quantization_args=quantization_args
+        )
+        module.register_module(f"{base_name}_observer", observer)
+
+
+def call_observer(module: Module, base_name: str, value: torch.Tensor):
+    """
+    Call a module's attached input/output observer using a provided value.
+    Update the module's scale and zp using the observer's return
+    values.
+
+    :param module: torch.nn.Module
+    :param base_name: substring used to fetch the observer, scales, and zp
+    :param value: torch.Tensor to be passed to the observer
+    """
+    observer = getattr(module, f"{base_name}_observer")
+    g_idx = getattr(module, "weight_g_idx", None)
+
+    updated_scale, updated_zero_point = observer(value, g_idx=g_idx)
+
+    # update scale and zero point
+    update_parameter_data(module, updated_scale, f"{base_name}_scale")
+    update_parameter_data(module, updated_zero_point, f"{base_name}_zero_point")
+
+
+def update_weight_zp_scale(module: Module):
```

Comment:
What was wrong with the name `set_module_for_calibration`? This new name seems like a misnomer since this function
1. only updates the quant_params of the weight
2. fake quantizes the weight
3. sets the status to calibration

Reply:
set_module_for_calibration is a misleading name as the function only impacts the weights, not activation quantization. 

This updates the zp and scales of the weights using the observers which is what the function name covers

Reply:
My mistake, the function does not quantize the weight
It does set the status to calibration though, it might be nice if the function name reflected this

Reply:
Either that or set the status to calibration outside of this function, which I think is preferred